dbms/src/Storages/Transaction/SchemaBuilder.cpp

```diff
@@ -331,22 +332,25 @@ void SchemaBuilder<Getter, NameMapper>::applyAlterPhysicalTable(DBInfoPtr db_inf
         return;
     }
 
-    std::stringstream ss;
-    ss << "Detected schema changes: " << name_mapper.debugCanonicalName(*db_info, *table_info) << ": ";
+    FmtBuffer fmt_buf;
+    fmt_buf.fmtAppend("Detected schema changes: {}: ", name_mapper.debugCanonicalName(*db_info, *table_info));
     for (const auto & schema_change : schema_changes)
         for (const auto & command : schema_change.first)
         {
             if (command.type == AlterCommand::ADD_COLUMN)
-                ss << "ADD COLUMN " << command.column_name << " " << command.data_type->getName() << ", ";
+                fmt_buf.fmtAppend("ADD COLUMN {} {},", command.column_name, command.data_type->getName());
             else if (command.type == AlterCommand::DROP_COLUMN)
-                ss << "DROP COLUMN " << command.column_name << ", ";
+                fmt_buf.fmtAppend("DROP COLUMN {}, ", command.data_type->getName());
             else if (command.type == AlterCommand::MODIFY_COLUMN)
-                ss << "MODIFY COLUMN " << command.column_name << " " << command.data_type->getName() << ", ";
+                fmt_buf.fmtAppend("MODIFY COLUMN {} {}, ", command.column_name, command.data_type->getName());
             else if (command.type == AlterCommand::RENAME_COLUMN)
-                ss << "RENAME COLUMN from " << command.column_name << " to " << command.new_column_name << ", ";
+                fmt_buf.fmtAppend("RENAME COLUMN from {} to {}, ", command.column_name, command.new_column_name);
         }
 
-    LOG_FMT_DEBUG(log, "{} : {}", __PRETTY_FUNCTION__, ss.str());
+    auto log_str = [&]() {
+        return fmt_buf.toString();
+    };
```

Comment:
the fmt construction cost is still there, you'd better wrap all of them into this lambda.

Reply:
Get.